core/src/retries.rs

```diff
@@ -22,22 +22,91 @@ fn find_io_error(original_error: &reqwest::Error) -> Option<&std::io::Error> {
     None
 }
 
-/// An [`ExponentialBackoff`] with parameters suitable for most HTTP requests. The parameters are
-/// copied from the parameters used in the GCP Go SDK[1].
+#[derive(Copy, Clone, Debug)]
+/// Extends the Backon ExponentialBackoff and ExponentialBuilder to add
+/// a `total_delay` field, which we use for deadlining in the collector's
+/// `poll_until_complete` method, as the underlying Backoff implementation
+/// does not expose that delay after it's set.
+pub struct ExponentialWithTotalDelayBuilder {
+    pub total_delay: Option<Duration>,
+    builder: ExponentialBuilder,
+}
+
+impl Default for ExponentialWithTotalDelayBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ExponentialWithTotalDelayBuilder {
+    pub const fn new() -> Self {
+        Self {
+            total_delay: None,
+            builder: ExponentialBuilder::new(),
+        }
+    }
+
+    pub const fn with_min_delay(mut self, d: Duration) -> Self {
+        self.builder = self.builder.with_min_delay(d);
+        self
+    }
+
+    pub const fn with_max_delay(mut self, d: Duration) -> Self {
+        self.builder = self.builder.with_max_delay(d);
+        self
+    }
+
+    pub const fn with_factor(mut self, f: f32) -> Self {
+        self.builder = self.builder.with_factor(f);
+        self
+    }
+
+    pub const fn with_max_times(mut self, t: usize) -> Self {
+        self.builder = self.builder.with_max_times(t);
+        self
+    }
+
+    pub const fn without_max_times(mut self) -> Self {
+        self.builder = self.builder.without_max_times();
+        self
+    }
+
+    pub const fn with_jitter(mut self) -> Self {
+        self.builder = self.builder.with_jitter();
+        self
+    }
+
+    pub fn with_total_delay(mut self, d: Option<Duration>) -> Self {
```

Comment:
Could this take `Duration` instead of `Option<Duration>`? Does anyone ever need to clear the total delay that's been set on one of these?

Reply:
I'd rather leave it as-is, to make these all consistent without having to use our composite type where it's unnecessary.